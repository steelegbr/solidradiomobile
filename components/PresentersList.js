/**
    Solid Radio Mobile App
    Copyright (C) 2020-2021 Marc Steele

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

/**
 * Presenter listings for a given station.
 */

import React from 'react';
import { connect } from 'react-redux';
import { Component } from 'react';
import { FlatList, View, StyleSheet, ActivityIndicator } from 'react-native';
import { List, Text } from 'react-native-paper';
import { loadStationPresenters } from '../reducers/actions';

class PresentersList extends Component {

    render() {

        const { loaded, presenters, navigation, stationName, loadStationPresenters } = this.props;

        if (loaded) {

            return (
                <FlatList
                    data={presenters}
                    renderItem={({ item }) => {

                        return (
                            <List.Item
                                title={item.name}
                                left={props => <List.Icon icon={{ uri: item.image }} />}
                                onPress={() => {
                                    navigation.push('PresenterDetail', {
                                        presenter: item
                                    });
                                }}
                            />
                        )

                    }}
                    keyExtractor={(item, index) => index.toString()}
                />
            );

        } else {

            // Trigger a load

            loadStationPresenters(stationName);

            // Show the loading screen

            return (
                <View>
                    <ActivityIndicator
                        style={styles.spinner}
                        size="large"
                    />
                    <Text
                        style={styles.text}
                    >
                        Loading top talent...
                    </Text>
                </View>
            );

        }



    }

}

const styles = StyleSheet.create({
    text: {
        flex: 0,
        width: "100%",
        fontSize: 20,
        fontWeight: "bold",
        textAlign: "center",
        padding: 20
    },
    spinner: {
        flex: 1,
        margin: 50
    },
    button: {
        margin: 20
    }
});

const mapStateToProps = (state, ownProps) => {

    // Work out if we have presenters or not

    const currentStation = state.presenters.currentStation;

    if (currentStation in state.presenters.stations) {
        return {
            loaded: true,
            presenters: state.presenters.stations[currentStation],
            navigation: ownProps.navigation,
            stationName: currentStation
        }
    } else {
        return {
            loaded: false,
            presenters: [],
            navigation: ownProps.navigation,
            stationName: currentStation
        }
    }

}

const mapDispatchToProps = (dispatch) => {
    return {
        loadStationPresenters: (stationName) => dispatch(loadStationPresenters(stationName))
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(PresentersList);