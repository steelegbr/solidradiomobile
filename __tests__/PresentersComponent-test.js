/**
    Solid Radio Mobile App
    Copyright (C) 2020-2021 Marc Steele

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

/**
 * Tests the presenter list component(s).
 */

import React from 'react';
import PresentersList from '../components/PresentersList';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

// Tests

it('renders', () => {

    // Arrange

    const mockStore = configureStore([]);
    const store = mockStore({
        stationNames: ['Foo FM', 'FooBar AM'],
        presenters: {
            currentStation: 'Foo FM',
            stations: {
                'Foo FM': [
                    {
                        name: 'Bob Agg',
                        image: 'bob_agg.png',
                        biography: 'Dangling round in the wrong places.'
                    },
                    {
                        name: 'Garry Baldy',
                        image: 'garry_baldy.gif',
                        biography: 'Sleak up top.'
                    },
                    {
                        name: 'Tessa Tickle',
                        image: 'tessa_tickle.jpg',
                        biography: 'Making you laugh!'
                    }
                ]
            }
        }
    });

    // Act 

    const presentersList = renderer.create(
        <Provider store={store}>
            <PresentersList />
        </Provider>
    ).toJSON();

    // Assert

    expect(presentersList).toMatchSnapshot();

});

it('renders-loading', () => {

    // Arrange

    const mockStore = configureStore([]);
    const store = mockStore({
        stationNames: ['Foo FM', 'FooBar AM'],
        presenters: {
            currentStation: 'Foo FM',
            stations: {}
        }
    });

    // Act 

    const presentersList = renderer.create(
        <Provider store={store}>
            <PresentersList />
        </Provider>
    ).toJSON();

    // Assert

    expect(presentersList).toMatchSnapshot();

});