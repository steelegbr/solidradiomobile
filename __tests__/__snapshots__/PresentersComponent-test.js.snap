// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "biography": "Dangling round in the wrong places.",
        "image": "bob_agg.png",
        "name": "Bob Agg",
      },
      Object {
        "biography": "Sleak up top.",
        "image": "garry_baldy.gif",
        "name": "Garry Baldy",
      },
      Object {
        "biography": "Making you laugh!",
        "image": "tessa_tickle.jpg",
        "name": "Tessa Tickle",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            false,
            Array [
              Object {
                "padding": 8,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "margin": 8,
                  "width": 40,
                },
                undefined,
              ]
            }
          >
            <Image
              accessibilityElementsHidden={true}
              importantForAccessibility="no-hide-descendants"
              source={
                Object {
                  "uri": "bob_agg.png",
                }
              }
              style={
                Array [
                  Object {
                    "transform": Array [
                      Object {
                        "scaleX": 1,
                      },
                    ],
                  },
                  Object {
                    "height": 24,
                    "resizeMode": "contain",
                    "tintColor": undefined,
                    "width": 24,
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "marginVertical": 6,
                  "paddingLeft": 8,
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Bob Agg
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            false,
            Array [
              Object {
                "padding": 8,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "margin": 8,
                  "width": 40,
                },
                undefined,
              ]
            }
          >
            <Image
              accessibilityElementsHidden={true}
              importantForAccessibility="no-hide-descendants"
              source={
                Object {
                  "uri": "garry_baldy.gif",
                }
              }
              style={
                Array [
                  Object {
                    "transform": Array [
                      Object {
                        "scaleX": 1,
                      },
                    ],
                  },
                  Object {
                    "height": 24,
                    "resizeMode": "contain",
                    "tintColor": undefined,
                    "width": 24,
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "marginVertical": 6,
                  "paddingLeft": 8,
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Garry Baldy
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            false,
            Array [
              Object {
                "padding": 8,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "margin": 8,
                  "width": 40,
                },
                undefined,
              ]
            }
          >
            <Image
              accessibilityElementsHidden={true}
              importantForAccessibility="no-hide-descendants"
              source={
                Object {
                  "uri": "tessa_tickle.jpg",
                }
              }
              style={
                Array [
                  Object {
                    "transform": Array [
                      Object {
                        "scaleX": 1,
                      },
                    ],
                  },
                  Object {
                    "height": 24,
                    "resizeMode": "contain",
                    "tintColor": undefined,
                    "width": 24,
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "marginVertical": 6,
                  "paddingLeft": 8,
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Tessa Tickle
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`renders-loading 1`] = `
<View>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "flex": 1,
        "margin": 50,
      }
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
          "textAlign": "left",
        },
        Object {
          "flex": 0,
          "fontSize": 20,
          "fontWeight": "bold",
          "padding": 20,
          "textAlign": "center",
          "width": "100%",
        },
      ]
    }
  >
    Loading top talent...
  </Text>
</View>
`;
